### ISSUE TITLE
NCHW conv2d on webgl backend returns incorrect result.

### ISSUE THREAD

**Issue by qjia7** [tensorflow/tfjs#6308]
Below test fails on webgl backend, while pass on cpu backend.
```
  it('x=[4,2,2] f=[2,2,4,4] s=1 d=1 p=same NCHW', async () => {
    const inputDepth = 4;
    const inputShape: [number, number, number] = [inputDepth, 2, 2];
    const outputDepth = 4;
    const fSize = 2;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NCHW';
    const dilation = 1;

    const x = tf.tensor3d(
        [1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4], inputShape);
    const w = tf.tensor4d(
        [
          3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1,
          1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5,
          0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0,
        ],
        [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([4, 2, 2]);
    expectArraysClose(
        resultData,
        [90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36]);
  });
```
Expected: `[90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36]`
Actual: `[90, 90, 90, 90, 54, 54, 54, 54, 63, 63, 63, 63, 36, 36, 36, 36]`

**Comment by Linchenn** [comment]
Found both CPU and WebGL could pass the following test for `NHWC` format:
```typescript
it('x=[2,2,4] f=[2,2,4,4] s=1 d=1 p=same NHWC', async () => {
    const inputDepth = 4;
    const inputShape: [number, number, number] = [2, 2, inputDepth];
    const outputDepth = 4;
    const fSize = 2;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NHWC';
    const dilation = 1;

    const x = tf.tensor3d(
        [1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4], inputShape);
    const w = tf.tensor4d(
        [
          3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1,
          1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5,
          0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0,
        ],
        [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 4]);
    expectArraysClose(
        resultData,
        [90, 90, 90, 90, 54, 54, 54, 54, 63, 63, 63, 63, 36, 36, 36, 36]);
  });
```
The result is similar with the wrong result of WebGL/`NHWC`.

Also compared the calling history between `NHWC` and `NCHW` of WebGL, found all arguments are correct before calling [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L45) (all cases I tested for this issue go into this if-branch `env().getBool('WEBGL_CONV_IM2COL') && x.shape[0] === 1`).

As a result, the issue could come from:
1. [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L45) missed format.
2. The [reshape](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L51) should consider the format.

**Comment by Linchenn** [comment]
Checked the implementation of [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L45), which mainly has 3 steps:
1. Use Im2ColPackedProgram to rearranges conv2d input. This step [considers](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/im2col_packed_gpu.ts#L75) the format and **the result is [compatible](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/im2col_packed_gpu.ts#L73) with `NHWC` format** (to compute the dot product with the filter, while the filter is always `NHWC` format).
2. Compute the dot product between the input and the filter.
3. Reshape the result.

This bug likely comes from the step3. For `NCHW` conv2d, the result of step2 is compatible with `NHWC` format, while the step3 does not permutate the result to be compatible with `NCHW` format before reshaping the result.

**Comment by Linchenn** [comment]
Checked the intermediate variable `im2Col`, the result of the step 1. Found the `NHWC` format and the `NCHW` format generate the same `im2ColReshaped`, while both step2 and step3 do not consider the format. **This proves the above assumption: step 3 is problematical**.

Specifically, both of the following tests have the same `backend().readSync(im2Col.dataId)`  as `[1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 2, 0, 4, 0, 2, 0, 4, 0, 2, 0, 4, 0, 2, 0, 4, 0, 3, 4, 0, 0, 3, 4, 0, 0, 3, 4, 0, 0, 3, 4, 0, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 0, 0]`:
```typescript
it('x=[4,2,2] f=[2,2,4,4] s=1 d=1 p=same NCHW', async () => {
    const inputDepth = 4;
    const inputShape: [number, number, number] = [inputDepth, 2, 2];
    const outputDepth = 4;
    const fSize = 2;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NCHW';
    const dilation = 1;

    const x = tf.tensor3d(
        [1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4], inputShape);
    const w = tf.tensor4d(
        [
          3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1,
          1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5,
          0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0,
        ],
        [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([4, 2, 2]);
    expectArraysClose(
        resultData,
        [90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36, 90, 54, 63, 36]);
  });

  it('x=[2,2,4] f=[2,2,4,4] s=1 d=1 p=same NHWC', async () => {
    const inputDepth = 4;
    const inputShape: [number, number, number] = [2, 2, inputDepth];
    const outputDepth = 4;
    const fSize = 2;
    const pad = 'same';
    const stride = 1;
    const dataFormat = 'NHWC';
    const dilation = 1;

    const x = tf.tensor3d(
        [1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4], inputShape);
    const w = tf.tensor4d(
        [
          3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1,
          1, 1, 5, 5, 5, 5, 0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5,
          0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 5, 5, 5, 5, 0, 0, 0, 0,
        ],
        [fSize, fSize, inputDepth, outputDepth]);

    const result = tf.conv2d(x, w, stride, pad, dataFormat, dilation);

    const resultData = await result.data();
    expect(result.shape).toEqual([2, 2, 4]);
    expectArraysClose(
        resultData,
        [90, 90, 90, 90, 54, 54, 54, 54, 63, 63, 63, 63, 36, 36, 36, 36]);
  });
```

**Comment by Linchenn** [comment]
Found there are [3 branches](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L39-L49) to compute Conv2d on WebGL.

The above comments are talking about [the second branch](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L44-L45), which uses Im2Row algorithm, and this branch is problematic.

Also found [the first branch](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L39-L43) is problematic too, because it does not consider the `NCHW` format.

**Comment by Linchenn** [comment]
In WebGL's [Conv2d_impl.ts](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts), both [conv2dByMatMul](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L44) and [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L184) take `bias` and `preluActivationWeights` as parameters. 

`bias` is supposed to be a 1-D tensor, so the dataFormat would not affect it. Added some tests in https://github.com/tensorflow/tfjs/pull/6340.

`preluActivationWeights`, as defined in [fusedConv2d](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-core/src/ops/fused/conv2d.ts#L91), the shape could be similar with the input, which means it could have multiple dimensions and `dataFormat` may affect it:

>  @param preluActivationWeights Tensor of prelu weights to be applied as part of a `prelu` activation, typically the same shape as `x`.

In conclusion, `preluActivationWeights` deserves more research to check the correctness for `NCHW` format in WebGL, but this is unrelated to this ISSUE, so I opened [a new one](https://github.com/tensorflow/tfjs/issues/6354) to track.
